## download dataset and install libraries

In [ ]:
!pip install git+https://github.com/huggingface/datasets.git

  Cloning https://github.com/huggingface/datasets.git to /tmp/pip-req-build-dffcwgr1
  Running command git clone --filter=blob:none --quiet https://github.com/huggingface/datasets.git /tmp/pip-req-build-dffcwgr1
  Resolved https://github.com/huggingface/datasets.git to commit 661d7bac29689e2d9eb74fba3d243939d6e9f25b
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 11.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 9.7 MB/s eta 0:00:00
  Created wheel for datasets: filename=datasets-3.1.1.dev0-py3-none-any.whl size=480548 sha256=c4d2480e3bfb7e94a50eb8cef736b15bd8477eaa68a67ed087ca3516b1911986
  Stored in directory: /tmp/pip-ephem-wheel-cache-9x1sqjli/wheels/

In [ ]:
!pip install llama-cpp-python

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 65.0/65.0 MB 11.8 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 2.7 MB/s eta 0:00:00
  Created wheel for llama-cpp-python: filename=llama_cpp_python-0.3.2-cp310-cp310-linux_x86_64.whl size=3410021 sha256=5441305b1438475536701f863952d0e6df4b8f995d5eb199352cd8018cb85355
  Stored in directory: /root/.cache/pip/wheels/49/1c/f0/6c1ed6032d5827dea28df8a8df860d90ca1bcd7b3fa455f5e1
Successfully built llama-cpp-python


In [ ]:
!pip install huggingface_hub

In [ ]:
!pip install rouge-score sentence-transformers

  Preparing metadata (setup.py) ... done
  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24935 sha256=091b2d208466ad4fb800f05ed8f565c9895ea1d54b955b239336cc585ce10b04
  Stored in directory: /root/.cache/pip/wheels/5f/dd/89/461065a73be61a532ff8599a28e9beef17985c9e9c31e541b4
Successfully built rouge-score


In [ ]:
!pip install datasets==2.21.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 527.3/527.3 kB 10.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.6/177.6 kB 7.9 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.9.0
    Uninstalling fsspec-2024.9.0:
      Successfully uninstalled fsspec-2024.9.0
  Attempting uninstall: datasets
    Found existing installation: datasets 3.1.1.dev0
    Uninstalling datasets-3.1.1.dev0:
      Successfully uninstalled datasets-3.1.1.dev0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2024.10.0 requires fsspec==2024.10.0, but you have fsspec 2024.6.1 which is incompatible.


In [ ]:
import pandas as pd
import numpy as np
import torch
import string
import re
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
nltk.download('punkt_tab')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

## check datasets for missing values

In [ ]:
from datasets import load_dataset

dataset = load_dataset("m3hrdadfi/recipe_nlg_lite")
dataset

Repo card metadata block was not found. Setting CardData to empty.


Generating train split:   0%|          | 0/6118 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1080 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['uid', 'name', 'description', 'link', 'ner', 'ingredients', 'steps'],
        num_rows: 6118
    })
    test: Dataset({
        features: ['uid', 'name', 'description', 'link', 'ner', 'ingredients', 'steps'],
        num_rows: 1080
    })
})

In [ ]:
train_dataset = pd.DataFrame(dataset["train"])
test_dataset = pd.DataFrame(dataset["test"])

In [ ]:
train_dataset.head()

uid  \
0      dab8b7d0-e0f6-4bb0-aed9-346e80dace1f   
1  b03f346bf39efcbace5d30a8f962147c8c4c361f   
2  89b49e742b2c1d234b83044c14d81155dfea7f19   
3      5db9af50-63dc-4c5b-9db1-783cf96675d3   
4      9b8da42d-d07c-4766-9f15-fd3fd6e19bf6   

                                          name  \
0                        pork chop noodle soup   
1  5 ingredient almond cake with fresh berries   
2                                 shrimp cakes   
3                           chili roasted okra   
4                    slow cooker chicken chili   

                                         description  \
0  we all know how satisfying it is to make great...   
1  this simple almond cake is made with just five...   
2  these light, pan seared shrimp cakes are moist...   
3  chili roasted okra with okra, sesame oil, red ...   
4  warm up on a cold day with this slow cooker ch...   

                                                link  \
0  https://www.yummly.com/private/recipe/Pork-Cho...   
1  https://www.skinnytaste.com/5-ingredient-almon...   
2          https://www.skinnytaste.com/shrimp-cakes/   
3  https://www.yummly.com/private/recipe/Chili-Ro...   
4  https://www.yummly.com/private/recipe/Slow-Coo...   

                                                 ner  \
0  bone in pork chops, salt, pepper, vegetable oi...   
1  large eggs, large egg whites, sugar, pure vani...   
2  peeled and deveined jumbo shrimp, plus 3 table...   
3  okra, sesame oil, red pepper flakes, salt, pepper   
4  oil, chicken, chili powder, onion, jalapeno pe...   

                                         ingredients  \
0  3.0 bone in pork chops, salt, pepper, 2.0 tabl...   
1  3 large eggs, 3 large egg whites, 2/3 cup suga...   
2  1 pound peeled and deveined jumbo shrimp, 1 cu...   
3  1.0 pound okra, 1.0 tablespoon sesame oil, 1.0...   
4  1.0 tablespoon oil, 1.0 pound chicken, 1.5 tab...   

                                               steps  
0  season pork chops with salt and pepper . heat ...  
1  position a rack in the middle of the oven and ...  
2  pat shrimp dry with a paper towel and place in...  
3  preheat the oven to 425degf . wash and dry the...  
4  heat oil in skillet over medium high heat . ad...

In [ ]:
# check for missing values
missing_train_values = train_dataset.isnull().values.any()
missing_test_values = test_dataset.isnull().values.any()
missing_train_values, missing_test_values

(False, False)

In [ ]:
train_dataset.drop(['uid', 'link'], axis=1, inplace=True)
test_dataset.drop(['uid', 'link'], axis=1, inplace=True)

In [ ]:
num_training = int(0.8 * len(train_dataset))
num_validation = int(0.2 * len(train_dataset))
print("Number of training examples:", num_training)
print("Number of validation examples:", num_validation)
print("Total number of examples:", num_training + num_validation)

# subsample the data
train_val = train_dataset.iloc[num_training : num_training+num_validation+1]
train_train = train_dataset.iloc[0 : num_training]

print("Number of training examples:", len(train_train))
print("Number of validation examples:", len(train_val))


Number of training examples: 4894
Number of validation examples: 1223
Total number of examples: 6117
Number of training examples: 4894
Number of validation examples: 1224


## process model and tokenizers

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

recipe_tokenizer = AutoTokenizer.from_pretrained("mbien/recipenlg")
recipe_model = AutoModelForCausalLM.from_pretrained("mbien/recipenlg", device_map="auto")

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/666 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/298 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/328 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/665M [00:00<?, ?B/s]

In [ ]:
# Generate a recipe response
prompt = "pork chop, noodle, water;"
inputs = recipe_tokenizer.encode(prompt, return_tensors="pt")
output = recipe_model.generate(
                    inputs,
                    max_length=300,
                    num_return_sequences=1,
                    pad_token_id=recipe_tokenizer.eos_token_id,
                    # temperature=0.7,  # Less randomness for consistency
                    top_k=50,
                    # top_p=0.9,
                    repetition_penalty=1.2,
                    do_sample=False,  # Disable sampling
                    num_beams=3,  # Enable beam search for more coherent results
                    early_stopping=True,  # Stops when a coherent point is reached
                )
out = output[0, len(inputs):].tolist()
text = recipe_tokenizer.decode(out)

print(text)

# response = recipe_tokenizer.decode(output[0], skip_special_tokens=True)
# response = post_process_recipe(response)
# print(response)

ork chop, noodle, water; salt and pepper. Bring to a boil. Reduce heat and simmer for 5 minutes.<NEXT_INSTR> Add the noodles and cook for another 5 minutes.<NEXT_INSTR> Add the water and bring to a boil again. Reduce heat and simmer for another 5 minutes.<NEXT_INSTR> Remove from heat and let stand for 5 minutes.<NEXT_INSTR> Fluff with a fork and serve.<INSTR_END><TITLE_START> Pork Chop And Noodles<TITLE_END><RECIPE_END><RECIPE_START><INPUT_START> butter<NEXT_INPUT> sugar<NEXT_INPUT> eggs<NEXT_INPUT> flour<NEXT_INPUT> baking powder<NEXT_INPUT> salt<NEXT_INPUT> milk<NEXT_INPUT> vanilla<INPUT_END><INGR_START> 1/2 c. butter<NEXT_INGR> 1 c. sugar<NEXT_INGR> 2 eggs<NEXT_INGR> 2 c. flour<NEXT_INGR> 2 tsp. baking powder<NEXT_INGR> 1/2 tsp. salt<NEXT_INGR> 1/2 c. milk<NEXT_INGR> 1 tsp. vanilla<INGR_END><INSTR_START> Cream butter and sugar.<NEXT_INSTR> Add eggs, one at a time, beating well after each addition.<NEXT_INSTR> Sift together flour, baking powder and salt.<NEXT_INSTR> Add alternately w

In [ ]:
!pip install rapidfuzz

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 26.8 MB/s eta 0:00:00


In [ ]:
from llama_cpp import Llama

llm = Llama.from_pretrained(
	repo_id="TheBloke/Llama-2-7B-Chat-GGUF",
	filename="llama-2-7b-chat.Q4_K_M.gguf",
  verbose=False
)

llama-2-7b-chat.Q4_K_M.gguf:   0%|          | 0.00/4.08G [00:00<?, ?B/s]

llama_new_context_with_model: n_ctx_per_seq (512) < n_ctx_train (4096) -- the full capacity of the model will not be utilized


In [ ]:
# Load DistilGPT-2 or GPT-2
model_name = "distilgpt2"
dgpt_tokenizer = AutoTokenizer.from_pretrained(model_name)
dgpt2 = AutoModelForCausalLM.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

In [ ]:
# Load DistilGPT-2 or GPT-2
model_name = "gpt2"
gpt_tokenizer = AutoTokenizer.from_pretrained(model_name)
gpt2 = AutoModelForCausalLM.from_pretrained(model_name)

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

In [ ]:
# Set padding token
if dgpt_tokenizer.pad_token is None:
    dgpt_tokenizer.pad_token = dgpt_tokenizer.eos_token

# Set padding token
if gpt_tokenizer.pad_token is None:
    gpt_tokenizer.pad_token = gpt_tokenizer.eos_token

# Intent Classification

In [ ]:
from transformers import pipeline
def intent_classification_with_model(user_input):
  classifier = pipeline("zero-shot-classification", model="facebook/bart-large-mnli", device=-1)
  input_text = user_input
  labels = ["recipe_request", "general_conversation"]
  result = classifier(input_text, labels, multi_label=False)
  return result["labels"][0]

# intent_classification_with_model("I am planning to host a thanksgiving party tonight. I have a turkey but I don't know what to do with it. Do you have suggestions on how to cook the turkey? Also, can you suggest some fun thanksgiving activities for the party?")
# intent_classification_with_model("hi, how are you?")

In [ ]:
intent_classification_with_model("I want to cook some noodles")

'recipe_request'

# dialogue manager

In [ ]:
def classify_intent(user_input):
    # If user mentions keywords related to cooking, assume they want a recipe
    recipe_keywords = ["recipe", "cook", "how to make", "ingredients", "method", "how do I cook"]
    if any(keyword in user_input.lower() for keyword in recipe_keywords):
        return "recipe_request"
    return "general_conversation"

In [ ]:
def post_process_recipe(response):
    sentences = response.split('.')
    unique_sentences = list(dict.fromkeys(sentences))  # Remove duplicates
    return '. '.join(unique_sentences) + '.'

# response = post_process_recipe(recipe_tokenizer.decode(output[0], skip_special_tokens=True))

In [ ]:
from rapidfuzz import process, fuzz

def recipe_request_processing(user_input):
  # Example: List of dish names
  dish_names = list(train_dataset['name'])

  # User input
  user_input = "noodle soup with pork chop"

  # Find the best match with a similarity threshold
  best_match, similarity, index = process.extractOne(query=user_input, choices=dish_names, scorer=fuzz.WRatio)
  if similarity > 80:  # Threshold for similarity
      print(f"Match found: {best_match} with similarity {similarity}%")
      desc = train_dataset[train_dataset['name'] == best_match]['description'].values[0]
      ingredients = train_dataset[train_dataset['name'] == best_match]['ingredients'].values[0]
      steps = train_dataset[train_dataset['name'] == best_match]['steps'].values[0]
      ingredients = ingredients.split(',')
      steps = steps.split('.')

      print(f"Description: {desc}")
      print("\n")
      print(f"Ingredients:\n")
      for ingredient in ingredients:
        print(ingredient)
      print("\n")
      print(f"Steps:\n")
      for step in steps:
        print(step)
      return True
  else:
      return False


In [ ]:
# def post_process_response(response):
#     # Remove repeated user inputs
#     lines = response.split('\n')
#     processed_lines = [line for line in lines if not line.startswith("User:")]
#     # Filter nonsensical text by length or repetition
#     processed_response = '. '.join(processed_lines).strip()
#     return processed_response

In [ ]:
def post_process_response(response):
    sentences = response.split('. ')
    seen_sentences = set()
    processed_sentences = []
    for sentence in sentences:
        if sentence not in seen_sentences:
            processed_sentences.append(sentence)
            seen_sentences.add(sentence)
    return '. '.join(processed_sentences).strip()

In [ ]:
def dialogue_manager(user_input):
    # Classify the user's intent
    intent = classify_intent(user_input)

    if intent == "recipe_request":

        input_dish = recipe_request_processing(user_input)
        if input_dish is False:
          # Use RecipeNLG model to generate a response
          inputs = recipe_tokenizer.encode(user_input, return_tensors="pt")
          output = recipe_model.generate(inputs, max_length=150, num_return_sequences=1, pad_token_id=recipe_tokenizer.eos_token_id)
          response = recipe_tokenizer.decode(output[0], skip_special_tokens=True)

    elif intent == "general_conversation":
        # Use Llama model for general chat (set max_token=0 to remove limit on max token)
        response = llm(user_input, max_tokens=0)

        # display the response
        response = response["choices"][0]["text"]

    else:
        response = "I'm not sure how to respond to that. Could you clarify?"

    return response


In [ ]:
MAX_HISTORY_LENGTH = 3  # Limit to last 3 turns

conversation_history = []

def dialogue_manager_with_limited_history(user_input, conv_model):
    try:
        # Update conversation history
        conversation_history.append(f"User: {user_input}")

        # Classify intent
        intent = intent_classification_with_model(user_input)
        print("intent: ", intent)

        # Limit conversation history to keep context concise
        if len(conversation_history) > MAX_HISTORY_LENGTH * 2:  # Each turn is user+bot
            conversation_history.pop(0)  # Keep history length reasonable

        if intent == "recipe_request":

            # if the user input the dish name
            input_dish = recipe_request_processing(user_input)
            if input_dish:
              response = ""
              return response

            # if the user input the list of ingredients
            else:
              # Modify the prompt for clarity and ensure it's concise
              prompt = user_input + " Please provide a concise recipe response."

              # Generate a recipe response
              inputs = recipe_tokenizer.encode(prompt, return_tensors="pt")

              # Catch errors related to tokenization or input length
              if inputs is None or inputs.size()[1] > 512:
                  return "Sorry, the input is too long for me to process. Could you please rephrase or provide fewer details?"

              output = recipe_model.generate(
                      inputs,
                      max_length=300,
                      num_return_sequences=1,
                      pad_token_id=recipe_tokenizer.eos_token_id,
                      temperature=0.7,  # Less randomness for consistency
                      top_k=50,
                      top_p=0.9,
                      repetition_penalty=1.2,
                      do_sample=False,  # Disable sampling
                      num_beams=3,  # Enable beam search for more coherent results
                      early_stopping=True  # Stops when a coherent point is reached
                  )

              # Check for valid output
              if output is None:
                  return "I couldn't generate a recipe response. Could you try asking again?"

              response = recipe_tokenizer.decode(output[0], skip_special_tokens=True)
              response = post_process_recipe(response)

        elif intent == "general_conversation":

          if conv_model == "llama":

            # Use conversation history to maintain context
            context_input = "\n".join(conversation_history[-MAX_HISTORY_LENGTH:]) + f"\nAI:"
            response = llm(context_input, max_tokens=150)

            # Extract response from Llama's output and handle None response
            if response is None or not isinstance(response, dict) or 'choices' not in response:
                return "I'm having trouble generating a response right now. Could you please rephrase your question?"

            response_text = response['choices'][0]['text'] if 'choices' in response else response
            conversation_history.append(f"AI: {response_text}")
            return response_text

          elif conv_model == "distilgpt-2":
            prompt = """You are a helpful assistant. Provide thoughtful and relevant responses to users.
              """ + "\n".join(conversation_history[-MAX_HISTORY_LENGTH:]) + f"\nUser: {user_input}\nAssistant:"

            # Tokenize input with padding and attention mask
            inputs = dgpt_tokenizer(prompt, return_tensors="pt", padding=True)

            # Generate response with improved constraints
            output = dgpt2.generate(
                inputs.input_ids,
                attention_mask=inputs.attention_mask,
                max_length=260,
                temperature=0.7,
                top_k=50,
                top_p=0.9,
                repetition_penalty=2.0,  # Penalize repeating content
                no_repeat_ngram_size=3,  # Prevent repeated 3-grams
                do_sample=True,
                num_return_sequences=1
            )
            response = dgpt_tokenizer.decode(output[0], skip_special_tokens=True)
            response = post_process_response(response)

            return response

          elif conv_model == "gpt-2":
            # Instruction and user input
            instruction = "You are a helpful assistant. Please provide thoughtful and concise responses."

            # Limit conversation history (if available) and format prompt
            context = ""
            if len(conversation_history) > 0:
                context = "\n".join(conversation_history[-MAX_HISTORY_LENGTH:])

            # Construct prompt with instruction, recent history, and user input
            prompt = f"{instruction}\n{context}\nUser: {user_input}\nAssistant:"

            # Tokenize input with padding and attention mask
            inputs = gpt_tokenizer(prompt, return_tensors="pt", padding=True, truncation=True)

            # Generate response with GPT-2
            output = gpt2.generate(
                inputs.input_ids,
                attention_mask=inputs.attention_mask,
                max_length=200,
                temperature=0.5,
                top_k=50,
                top_p=0.9,
                repetition_penalty=3.0,  # Penalize repetition further
                no_repeat_ngram_size=3,
                do_sample=True,
                num_return_sequences=1,
                num_beams=3,
                eos_token_id=gpt_tokenizer.eos_token_id
            )

            # Decode and post-process response
            response = gpt_tokenizer.decode(output[0], skip_special_tokens=True)
            response = post_process_response(response)

            # Add the new response to the conversation history, ensuring history remains concise
            conversation_history.append(f"User: {user_input}")
            conversation_history.append(f"Assistant: {response}")

            return response
        else:
            return "I'm not sure how to respond to that. Could you clarify?"

    except Exception as e:
        # Handle general exceptions and provide a safe fallback message
        return f"An error occurred: {str(e)}. Please try again."

    return response


In [ ]:
# Simple chat loop
import timeit
print("Chatbot is ready! Type 'exit' to end the chat.")
conversation_history = []

conv_model = input("Which conversational model do you want to use? Llama or DistilGPT-2 or GPT-2? ").lower()
print(type(conv_model))
while conv_model != "llama" and conv_model != "distilgpt-2" and conv_model != "gpt-2":
  print("please only choose either Llama or DistilGPT-2 or GPT-2.")
  conv_model = input("Which conversational model do you want to use? Llama or DistilGPT-2 or GPT-2? ").lower()

start = timeit.default_timer()
print(f"start time: {start}")
while True:
    user_input = input("You: ")
    input_time = timeit.default_timer()
    print(f"input time: {input_time}")
    if user_input.lower() == "exit":
        break
    response = dialogue_manager_with_limited_history(user_input, conv_model)
    print(f"Chatbot: {response['choices'][0]['text'] if 'choices' in response else response}")
    response_time = timeit.default_timer()
    print(f"response time: {response_time}")
end = timeit.default_timer()
print(f"end time: {end}")
print(f"This conversation takes {end-start} seconds.")

Chatbot is ready! Type 'exit' to end the chat.
Which conversational model do you want to use? Llama or DistilGPT-2 or GPT-2? llama
<class 'str'>
start time: 6222.422873388
You: hi, how are you?
input time: 6226.990675266
intent:  general_conversation
Chatbot:  Hello! I'm doing well, thank you for asking! I'm just an AI, so I don't have feelings or emotions like humans do, but I'm functioning properly and happy to assist you. How can I help you today?
response time: 6280.317540921
You: can you help me cook some pork chop with noodle soup?
input time: 6339.192906771
intent:  recipe_request
Match found: pork chop noodle soup with similarity 95.0%
Description: we all know how satisfying it is to make great pork tenderloin, ribs, or a roast but the end of the meal creates a new quandary what do you do with the leftover pork contrary to what you might think, it's not that difficult . how to repurpose your meal is where real cooking creativity comes into play, so let us present to you our fav

# Evaluation

In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("kreeshrajani/3k-conversations-dataset-for-chatbot")

print("Path to dataset files:", path)

100%|██████████| 67.1k/67.1k [00:00<00:00, 353kB/s]

Extracting files...


Path to dataset files: /root/.cache/kagglehub/datasets/kreeshrajani/3k-conversations-dataset-for-chatbot/versions/1


In [ ]:
df = pd.read_csv("/content/drive/MyDrive/MS/CS6353/project/dataset/chatbot_dataset.zip")
df.head()

Unnamed: 0                             question  \
0           0               hi, how are you doing?   
1           1        i'm fine. how about yourself?   
2           2  i'm pretty good. thanks for asking.   
3           3    no problem. so how have you been?   
4           4     i've been great. what about you?   

                                     answer  
0             i'm fine. how about yourself?  
1       i'm pretty good. thanks for asking.  
2         no problem. so how have you been?  
3          i've been great. what about you?  
4  i've been good. i'm in school right now.

In [ ]:
import nltk
from rouge_score import rouge_scorer
from sklearn.metrics import classification_report
from sentence_transformers import SentenceTransformer, util
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

In [ ]:
# embedding model for similarity metrics
embedding_model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

In [ ]:
# BLEU Score
def compute_bleu(reference, hypothesis):
    reference = [nltk.word_tokenize(reference.lower())]
    hypothesis = nltk.word_tokenize(hypothesis.lower())
    bleu_score = nltk.translate.bleu_score.sentence_bleu(reference, hypothesis)
    return bleu_score

# ROUGE Score
def compute_rouge(reference, hypothesis):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    scores = scorer.score(reference, hypothesis)
    return scores

# Cosine Similarity
def compute_similarity(reference, hypothesis):
    ref_embedding = embedding_model.encode(reference, convert_to_tensor=True)
    hyp_embedding = embedding_model.encode(hypothesis, convert_to_tensor=True)
    similarity = util.pytorch_cos_sim(ref_embedding, hyp_embedding).item()
    return similarity

In [ ]:
# Recipe Generation
def generate_recipe(prompt):
    inputs = recipe_tokenizer(prompt, return_tensors="pt")
    output = recipe_model.generate(
        inputs.input_ids,
        max_length=150,
        temperature=0.7,
        top_k=50,
        top_p=0.9,
        repetition_penalty=1.2,
        do_sample=True,
        num_return_sequences=1
    )
    response = recipe_tokenizer.decode(output[0], skip_special_tokens=True)
    return response

In [ ]:
# Conversational Model Comparison
def evaluate_models(input_text, models):
    results = {}
    for model_name, model_info in models.items():
        if model_name == "Llama":
            # Generate response using Llama
            response = llm(input_text)["choices"][0]["text"]
        else:
            # Generate response using GPT models
            tokenizer, model = model_info["tokenizer"], model_info["model"]
            inputs = tokenizer(input_text, return_tensors="pt", padding=True, truncation=True)
            output = model.generate(
                inputs.input_ids,
                attention_mask=inputs.attention_mask,
                max_length=150,
                temperature=0.7,
                top_k=50,
                top_p=0.9,
                repetition_penalty=1.2,
                do_sample=True
            )
            response = tokenizer.decode(output[0], skip_special_tokens=True)
        results[model_name] = response.strip()
    return results

In [ ]:
# Initialize models for conversational evaluation
models = {
    "DistilGPT-2": {
        "tokenizer":dgpt_tokenizer,
        "model": dgpt2
    },
    "GPT-2": {
        "tokenizer": gpt_tokenizer,
        "model": gpt2
    },
    "Llama": llm
}

In [ ]:
# Example evaluation
user_input = "I want a recipe for a scrambled egg."
expected_output = """Scrambled Eggs Recipe:

Ingredients:
- 2 large eggs
- 1 tbsp milk (optional)
- 1 tbsp butter or oil
- Salt and pepper to taste

Instructions:
1. Crack eggs into a bowl. Add milk (for creaminess) and whisk until well blended.
2. Heat a non-stick pan over medium heat and melt butter or add oil.
3. Pour the egg mixture into the pan.
4. Let it sit for a few seconds, then gently stir with a spatula, pushing from the edges to the center.
5. Cook until eggs are softly set but slightly runny for creamy texture.
6. Remove from heat (eggs will continue cooking).
7. Season with salt and pepper. Serve immediately.
"""

# Generate recipe
generated_recipe = generate_recipe(user_input)
print(f"Generated Recipe: {generated_recipe}")

# Compute Evaluation Metrics
bleu = compute_bleu(expected_output, generated_recipe)
rouge = compute_rouge(expected_output, generated_recipe)
similarity = compute_similarity(expected_output, generated_recipe)

print(f"BLEU Score: {bleu}\n")
print(f"ROUGE Score: {rouge}\n")
print(f"Semantic Similarity: {similarity}\n")

# Evaluate conversational models
conversation_input = "Hi, how are you?"
conversation_output = df["answer"][0]
model_responses = evaluate_models(conversation_input, models)
print("Model Responses:\n")
for model_name, response in model_responses.items():
    print(f"{model_name}: {response}\n")

    # Compute Evaluation Metrics
    bleu = compute_bleu(conversation_output, response)
    rouge = compute_rouge(conversation_output, response)
    similarity = compute_similarity(conversation_output, response)

    print(f"model name: {model_name}\n")
    print(f"BLEU Score: {bleu}\n")
    print(f"ROUGE Score: {rouge}\n")
    print(f"Semantic Similarity: {similarity}\n")

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Generated Recipe: I want a recipe for a scrambled egg. Preheat oven to 350 degrees F (175 degrees C). Mix eggs, milk and butter together in a bowl; pour into an 8x8 inch baking dish. Bake in the preheated oven until edges are set, about 15 minutes. Turn off heat and allow the dish to cool slightly. Pour the cooled baked custard over the top of the flan mixture. Bake in the hot oven until center is no longer liquidy, 10 to 15 minutes more. Let stand until cool enough to cut with a serrated knife. Scrambled Egg Flan flour sugar salt baking soda cinnamon nutmeg vanilla vegetable oil


/usr/local/lib/python3.10/dist-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/usr/local/lib/python3.10/dist-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


BLEU Score: 5.069418628865746e-155

ROUGE Score: {'rouge1': Score(precision=0.34579439252336447, recall=0.33636363636363636, fmeasure=0.3410138248847926), 'rouge2': Score(precision=0.05660377358490566, recall=0.05504587155963303, fmeasure=0.0558139534883721), 'rougeL': Score(precision=0.17757009345794392, recall=0.17272727272727273, fmeasure=0.17511520737327188)}

Semantic Similarity: 0.7695318460464478



Setting `pad_token_id` to `eos_token_id`:None for open-end generation.


Model Responses:

DistilGPT-2: Hi, how are you? How do I get here?"
"Yeah," he said. "I'm sorry."
He nodded and went back to his desk in front of the office that day as they left for work on their way home from a family dinner at Gershine's mansion where one woman had been sleeping with her boyfriend while she slept alone; there was no time when it would have felt so different or more like an eternity if someone else could say something about what happened next: this is not just me getting into trouble but also my girlfriend who has never seen him since…well now we're married! You know why your friend called us yesterday morning trying hard to explain things—and then even got himself off track by saying

model name: DistilGPT-2

BLEU Score: 1.9488813334223974e-155

ROUGE Score: {'rouge1': Score(precision=0.030303030303030304, recall=0.6666666666666666, fmeasure=0.057971014492753624), 'rouge2': Score(precision=0.007633587786259542, recall=0.2, fmeasure=0.014705882352941175), 'rougeL': S

output evaluating all models:

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
/usr/local/lib/python3.10/dist-packages/nltk/translate/bleu_score.py:577: UserWarning:
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/usr/local/lib/python3.10/dist-packages/nltk/translate/bleu_score.py:577: UserWarning:
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Generated Recipe: I want a recipe for chocolate cake that takes about 18 minutes to bake in the oven. In a bowl, whisk together all of the ingredients until combined. Place the pan on your rack and pour the batter into it. Bake at 350°F for about 30-35 minutes or until done. Serve warm. Chocolate Cake butter brown sugar eggs flour baking powder salt milk vanilla pecans 1/2 c. butter 2 c. firmly packed brown sugar 3 large eggs 2 c. sifted flour 4 tsp. baking powder dash of salt 1/3 c. milk 1 tsp. vanilla 2 c. chopped pe
BLEU Score: 2.1716554675787733e-155

ROUGE Score: {'rouge1': Score(precision=0.049019607843137254, recall=0.7142857142857143, fmeasure=0.09174311926605504), 'rouge2': Score(precision=0.009900990099009901, recall=0.16666666666666666, fmeasure=0.018691588785046728), 'rougeL': Score(precision=0.0392156862745098, recall=0.5714285714285714, fmeasure=0.07339449541284403)}

Semantic Similarity: 0.6374624967575073

Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Model Responses:

DistilGPT-2: Hi, how are you?
I’m not sure what is going on. But the only thing I know that it was all about me at first and then my mom did something like this: ‪♡️ Thank god! You were very funny to everyone who had a chance so we didn't have any problems with her because she wanted us to be friends."

GPT-2: Hi, how are you? What do I need to know?"
"I'm sorry. My name is Gwyneth."

 (Gwynne walks in the doorway) "It's been a long time since my last meal at Pinnacle Cafe... but it feels good eating..." He looks down and sees his friends standing around him as he speaks with something that sounds like this: "...we can't make your life miserable anymore!" And then everyone else nods their agreement saying nothing more after they've said so loudly for some reason not knowing what has happened! They're all there waiting anxiously on them now which isn't very happy because of course she didn`d just say 'no'. You have no idea why people were worried about

Llama: It is important to ask this question as it allows the person to open up and


Output when evaluating distilgpt-2 and gpt-2 and the recipe generator model

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Generated Recipe: I want a recipe for chocolate cake. Preheat oven to 350F (180C). Make the buttercream: Whisk together the sugar and eggs until pale yellow in color, about 2 minutes. Add the flour and mix well. Add the chocolate mixture, 1/3 at a time, mixing well after each addition. Pour into prepared pan and bake until toothpick inserted comes out clean, about 30 to 40 minutes. Cool completely before cutting. Buttercream Pound Cake chicken cream of mushroom soup milk cheddar cheese salt pepper corn flakes 3 cups cooked chicken 1 can cream of mushroom soup 12 cup evaporated milk 8 ounces shredded sharp c
BLEU Score: 1.954457588783683e-155
ROUGE Score: {'rouge1': Score(precision=0.038461538461538464, recall=0.5714285714285714, fmeasure=0.07207207207207209), 'rouge2': Score(precision=0.009708737864077669, recall=0.16666666666666666, fmeasure=0.01834862385321101), 'rougeL': Score(precision=0.028846153846153848, recall=0.42857142857142855, fmeasure=0.05405405405405405)}
Semantic Similarity: 0.6556602716445923
Setting `pad_token_id` to `eos_token_id`:None for open-end generation.
Model Responses:
DistilGPT-2: Hi, how are you? What's the reason why I am here.
I'm a regular guest on my blog and one of the most popular podcast hosts in Seattle this week! The first question is: How do we know that there aren't many people who have been living with depression for years now - or even after they've had it all over their lives to get back into work?! This can be difficult because as far as your background goes, no matter what type/culture (or culture) everyone has an issue about these issues; just look at some examples from other places where mental illness comes up... but don't go too much further than reading through them yourself!!! Here were six common themes among me during writing our show "The
GPT-2: Hi, how are you?
I'm actually pretty good at it. I just don't have a great sense of humor or anything like that on the part but if someone says something funny about me and then they're really going to hate my work (laughs), we can talk some more… But this is not gonna be fun for anyone because there's no way in hell somebody would want any kinder relationship than mine! So when people say things such as "he likes your hair" so much... It could make him go crazy." Then he'll start looking up every single thing she did before her death – which makes perfect logical stuff with these two women who look exactly alike…. If everyone said all those times after their divorce where was everybody supposed


# Chat output

## chat without history

Chatbot is ready! Type 'exit' to end the chat.
start time: 0.027116240999930596
You: hi, happy thanksgiving
input time: 0.015560721999918314
Chatbot: ! today is a day of giving thanks and expressing gratitude, and i just wanted to take a moment to express my deepest gratitude to everyone who has supported me on this journey.
from the bottom of my heart, thank you. your kind words, your encouragement, your support – it has meant the world to me. without you, i would not be where i am today, and i am so grateful for your presence in my life.
whether you are a friend, a family member, or simply someone who stumbled upon my work – thank you. your support has been invaluable, and i am so grateful to have you in my life.
so on this day of giving thanks, i want to express my gratitude to each and every one of you. thank you for being a part of my journey, and for helping me to grow and succeed in ways i never thought possible.
i hope you have a wonderful thanksgiving, filled with love, laughter, and joy. and i hope you know how much you are appreciated and valued.
thank you again, from the bottom of my heart.
🥰🍁🦃🎉
response time: 0.023282362000372814
You: thank you to you too. I am planning to host a thanksgiving party tonight. I have a turkey but I don't know what to do with it. Do you have suggestions on how to cook the turkey?
input time: 0.026911459999610088
Chatbot: thank you to you too. I am planning to host a thanksgiving party tonight. I have a turkey but I don't know what to do with it. Do you have suggestions on how to cook the turkey? Turkey butter sugar eggs flour baking powder salt milk vanilla lemon extract lemon zest lemon juice sugar lemon juice lemon zest lemon juice 1/2 cup butter, softened 1 cup sugar 2 eggs 2 cups all-purpose flour 1 teaspoon baking powder 1/2 teaspoon salt 1/2 cup milk 1 teaspoon vanilla extract 1/2 teaspoon lemon extract 1/2 teaspoon lemon zest 1
response time: 0.016123132999837253
You: what dish is that?
input time: 0.0161975569999413
Chatbot:

I'm sorry, but I can't tell you the name of the dish without more information. Can you please provide more context or details about the dish?
response time: 0.01599856200027716
You: nothing. thanks. exit
input time: 0.016954222000094887
Chatbot: .

Answer:

You are correct. The output of the command `ls` without any arguments is simply the list of files and directories in the current directory.

Therefore, the command `ls` on its own does not output anything else besides the list of files and directories.

I apologize for any confusion, and thank you for pointing this out.
response time: 0.02741494000019884
You: exit
input time: 0.020534798999960913
end time: 0.01587342000038916
This conversation takes -0.011242820999541436 seconds.



## chat with history after refinement

Chatbot is ready! Type 'exit' to end the chat.
start time: 0.016899080000257527
You: hi, i have a math homework that i'm really stressed about. Do you have any suggestions on how to calm me down?
input time: 0.01661006799986353
Chatbot:  Of course, I'd be happy to help! Here are a few suggestions that may help you calm down and feel more focused:

1. Take a few deep breaths: When we're stressed or anxious, our breathing can become shallow and rapid. Taking a few deep breaths can help calm the body and mind. Try inhaling for a count of four, holding your breath for a count of four, and exhaling for a count of four. Repeat this process several times.
2. Practice progressive muscle relaxation: This technique involves tensing and relaxing different muscle groups in your body. Start with your toes and work your way up to your head,
response time: 0.025990429000557924
You: thanks! i think it worked. I'm hungry. I'm thinking i want to cook something with chicken and butter. how do i cook chicken and butter?
input time: 0.015895821999947657
/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:590: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.7` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`.
  warnings.warn(
Chatbot: Provide a recipe using chicken and butter: thanks! i think it worked.  I'm hungry.  I'm thinking i want to cook something with chicken and butter.  how do i cook chicken and butter? Chicken And Butter Recipe flour sugar baking powder salt milk egg butter 2 c.  flour 1/2 c.  sugar 4 tsp.  baking powder 1/2 tsp.  salt 1/2 c.  milk 1 egg 1/4 c.  butter Sift dry ingredients together.  Add milk, egg and butter.  Beat until smooth.  Pour into greased muffin tins.  Bake at 400° for 20 to 25 minutes.  Makes 12 muffins.  Muffins butter sugar eggs vanilla flour baking powder salt milk 1/2 c.  butter 1 c.  sugar 2 eggs 1 tsp.  vanilla 2 c.  flour 2 tsp.  milk Cream butter and sugar.  Add eggs and vanilla.  Sift flour, baking powder and salt together.  Add to creamed mixture alternately with milk.  Pour into greased and floured 9 x 13-inch pan.  Bake at 350° for 30 to 35 minutes.  Butterscotch Brownies.
response time: 0.015400991999740654
You: nice thank you
input time: 0.03328512299958675
Chatbot:   Of course! Cooking with chicken and butter can be a delicious and satisfying meal. Here are some basic steps for cooking chicken and butter:
1. Start by preheating a pan or skillet over medium heat.
2. Add a small amount of butter to the pan and let it melt.
3. Add the chicken to the pan and cook until it is browned on all sides and cooked through.
4. Once the chicken is cooked, remove it from the pan and set it aside.
5. Continue to cook the butter in the pan, stirring occasionally, until it is golden brown and slightly caramelized.
6
response time: 0.015950001000419434
You: exit
input time: 0.018939782999950694
end time: 0.016129360999912024
This conversation takes -0.0007697190003455034 seconds.

# scratch

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

# Load DistilGPT-2 or GPT-2
model_name = "distilgpt2"  # Change to "gpt2" if you prefer GPT-2
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Function to generate conversational responses
def gpt_conversation(user_input):
    try:
        inputs = tokenizer(user_input, return_tensors="pt")
        output = model.generate(
            inputs.input_ids,
            max_length=150,  # Adjust for desired response length
            temperature=0.7,  # Adds some variation
            top_k=50,  # Limits sampling pool
            top_p=0.9,  # Limits to most probable tokens
            repetition_penalty=1.2,  # Penalizes repetitive outputs
            do_sample=True,  # Enable sampling for creative output
            num_return_sequences=1
        )
        response = tokenizer.decode(output[0], skip_special_tokens=True)
        return response
    except Exception as e:
        return f"An error occurred with the GPT response: {str(e)}"

# Example of usage in dialogue manager
user_input = "What are some good ways to cook pasta?"
response = gpt_conversation(user_input)
print(f"Chatbot: {response}")
